docs/RFCS/stateless_replica_relocation.md

```diff
@@ -0,0 +1,232 @@
+for selecting an appropriate store for the complementary operation.
+
+It is assumed that this level of specificity is simply not needed for any
+relocation operations; there is no currently apparent use case where a move
+between a specific pair of stores is needed.
+
+Even if this was necessary, it might be possible to express by manipulating the
+factors behind the individual stateless decisions.
+
+### Thrashing of complementary operation
+Because there is no relocation state, the possibilty of "thrashing" is introduced. For
+example: 
+
+1. Rebalance operation adds a new replica to the range.
+2. The replicate queue picks up the range and detects the need to remove a
+   replica; however, it decides to remove the replica that was just added.
+
+This is possible if the rebalance's criteria for new replica selection are
+sufficiently different from the replicate queue's selection criteria for
+removing a replica.
+
+To reduce this risk, there must be sufficient agreement in the criteria between
+the operations; a Rebalance operation should avoid adding a new replica if
+there's a realistic chance that the replicate queue will immediately remove it.
+
+This can realized by adding a "buffer" zone between the different criteria; that
+is, when selecting a replica to add, the new replica should be *significantly*
+removed from the criteria for removing a replica, thus reducing the chances that
+it will be selected.
+
+When properly tuned to avoid thrashing, the stateless nature could instead be
+considered a positive because it can respond to changes during the interim;
+consider if, due to a delay, the relative health of nodes changes and the
+originally selected replica is no longer a good option.  The stateless system
+will respond correctly to this situation.
+
+### Delay in complementary operation
+By separating the two operations, we are possibly delaying the application of
+the second operation. For example, the replicate queue could be very busy, or an
+untimely crash could result in the range being under- or over-replicated without
+being in the replicate queue on any store.
+
+However, many of these concerns are allayed by the existence of repair; if the
+node goes down, the repair system will add the range to the replicate queue on
+another store.
+
+Even in an exotic failure scenario, the stateless design will eventually detect
+the replication anomaly through the normal operation of the replica scanner.
+
+### Lack of non-voting replicas
+Some types of relocation are fragile without non-voting member replicas.
+
+When initially creating a replica, it is very far behind the current state of
+the range and thus needs to receive a snapshot. It may take some time before the
+range fully catches up and can take place in quorum commits.
+
+However, without non-voting replicas we have no choice but to add the new
+replica as a full member, thus changing the quorum requirements of the group. In
```

Comment:
There is a less-intrusive alternative to non-voting replicas: We could manually generate a `MsgSnapshot` and send it to the new node (waiting for a response) before starting the transaction to formally add the replica. Then the window in which the quorum is jeopardized would be limited to the time needed to send the tail of the logs since the snapshot was generated.